src/exchanges/bitfinex-client.js

```diff
@@ -207,9 +234,79 @@ class BitfinexClient extends BasicClient {
   }
 
   _onTradeMessage(msg, market) {
-    // example msg: [ 359491, 'tu', [ 560287312, 1609712228656, 0.005, 33432 ], 6 ]
-    let [id, unix, amount, price] = msg[2];
+    const timestampMs = msg[3];
+    if (msg[1] === "hb") {
+      const sequenceId = Number(msg[2]);
+      if (this.enableEmptyHeartbeatEvents === false) return;
+      // handle heartbeat by emitting empty update w/sequenceId.
+      // example trade heartbeat msg: [ 198655, 'hb', 3, 1610920929093 ]
+      let trade = new Trade({
+        exchange: "Bitfinex",
+        base: market.base,
+        quote: market.quote,
+        timestamp: timestampMs,
+        sequenceId,
+      });
+      this.emit("trade", trade, market);
+      return;
+    }
+    if (Array.isArray(msg[1])) {
+      const sequenceId = Number(msg[2]);
+      // trade snapshot example msg:
+      /*
+      [
+        CHANNEL_ID,
+        [
+          [
+            ID,
+            MTS,
+            AMOUNT,
+            PRICE
+          ],
+          ...
+        ],
+        sequenceId,
+        timestampMs
+      ]
+      */
+      msg[1].forEach(thisTrade => {
```

Comment:
nit: avoid `forEach`, it's slower than `for...of`

Reply:
is that still true? either way I'm happy to make the change just curious. I know JS engine vendors are constantly making forEach faster since it's so commonly used. I ran a simple test with this benchmarking tool http://jsbench.github.io/ and showed them dead on equal on Chrome on Linux

I just did these side by side for the benchmark

```js
// for of
// set up an array to iterate over
let sourceArr = [];
let foo = null;
for (let i = 0; i < 10000; i++) {
    sourceArr.push(i);
}

for (el of sourceArr) {
    // something nonsensical to do for the test
   foo = el + 1; 
}
console.log('for of done, foo=', foo)
```

```js
// for each
// set up an array to iterate over
let sourceArr = [];
let foo = null;
for (let i = 0; i < 10000; i++) {
    sourceArr.push(i);
}

sourceArr.forEach((el) => {
    // something nonsensical to do for the test
   foo = el + 1; 
});
console.log('foreach done, foo=', foo)
```

Reply:
Yeah, definitely going to depend on the engine/version you're using (Firefox on Linux is ~75% faster). This lib avoids them in favor of `for..of`.  To be added to the list of eslint rules haha